### PR TITLE
State machines

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Adds a section about state machines: what to use them for, and how to implement one. Thanks!

### CODE REVIEW

File: content/guides/state-machines.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+function stateMachine (transitions, currentState, transitionName) {
+  var newState = transitions[currentState][transitionName]
+  return newState
+}
+```
+
+Or in a more compact notation:
+
+```js
+var stateMachine => (t, c, n) => t[c][n]
```
Comment (bendrucker) [inline content/guides/state-machines.md]: first `=>` should be `=`

File: content/guides/state-machines.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+- Parallel state machines.
+- Nested state machines.
+
+We'll leave this as an exercise up to the reader.
+
+## Nanostate
+If you're looking for a solid state machine implementation, check out
+[choojs/nanostate](https://github.com/choojs/nanostate). It's similar to the
+state machine we implemented in the section above, but adds event hooks, clean
+error messages and more.
```
Comment (bendrucker) [inline content/guides/state-machines.md]: oxford comma? (, and more)

**Comment by bendrucker** [comment]
Super well done, caught some typos for ya!

**Comment by yoshuawuyts** [comment]
Pushed fixes! Thanks @bendrucker :sparkles:

File: content/guides/state-machines.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# state machines
+If you work in software long enough, you'll have to deal with legacy codebases.
+but you'll also have had the chance to start projects from scratch. And
+eventually see your project become a legacy codebase itself.
```
Comment (ghost) [inline content/guides/state-machines.md]: `eventually you'll see [...]`

File: content/guides/state-machines.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+are all based on a timer.
+
+```txt
+green  => { timer: orange }
+orange => { timer: red }
+red    => { timer: green }
+```
+
+And that's the basics. From here we could go on to add more states and
+transitions, expanding the graph.
```
Comment (ghost) [inline content/guides/state-machines.md:52]: graph here introduces a concept that previously hasnt been talked about, might be good to add a representation of the state machine as a graph, visually

File: content/guides/state-machines.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+## State machines in JavaScript
+Let's implement the traffic light example in JavaScript. In order for this to
+work, we'll need to implement:
+- save the states & transitions in an Object
```
Comment (ghost) [inline content/guides/state-machines.md]: `saving`

File: content/guides/state-machines.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+- the core state machine algorithm
+- create a small, stateful class to hold the state
```
Comment (ghost) [inline content/guides/state-machines.md]: `creating`

File: content/guides/state-machines.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+- combine all of these to form the complete state machine.
```
Comment (ghost) [inline content/guides/state-machines.md]: `combining`

File: content/guides/state-machines.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+Or in a more compact notation:
+
+```js
+var stateMachine = (t, c, n) => t[c][n]
+```
+
+### Class
+While the state machine algorithm is quite simple, it requires us to keep track
+of what the current state is. This means that statemachines themselves are
```
Comment (ghost) [inline content/guides/state-machines.md]: state machines

File: content/guides/state-machines.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+From here there are a few more features that could be added, such as:
+- Event hooks to trigger events when certain transitions happen.
+- Combining several state machines into a larger one.
+- Automatic traversal of state machines using pathfinding algorithms.
+- Parallel state machines.
+- Nested state machines.
+
+We'll leave this as an exercise up to the reader.
+
```
Comment (ghost) [inline content/guides/state-machines.md:164]: imo this is missing usage examples in user interface design, as hinted at during the first section

**Review by ghost — CHANGES_REQUESTED** [review]

File: content/guides/state-machines.md (hunk 4 above)
Comment (yoshuawuyts) [reply content/guides/state-machines.md:52]: I agree, but don't have any graph tools around right now. If someone has a graph that'd be neat!

File: content/guides/state-machines.md (hunk 9 above)
Comment (yoshuawuyts) [reply content/guides/state-machines.md:164]: aye, yeah fair

**Comment by yoshuawuyts** [comment]
Made an issue for the last unsolved comment. Merging now, will fix the remaining comments later! :sparkles: 

V. excited to have some of these sections up! :grin:

**Pull request merged**